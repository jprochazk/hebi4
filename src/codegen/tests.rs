use std::fs::read_to_string;

use super::emit;
use crate::{parser::parse, token::tokenize};

uitest::uitest!("tests/inputs/codegen/*.hi", "codegen", |path| {
    let input = read_to_string(path).unwrap();
    let tokens = tokenize(&input);
    let ast = match parse(&tokens) {
        Ok(ast) => ast,
        Err(err) => {
            panic!("{}", err.render(&input));
        }
    };
    let (snapshot, failure) = match emit(&ast) {
        Ok(chunk) => (format!("OK\n{}", chunk.disasm(&input)), false),
        Err(err) => (format!("ERROR\n{}", err.render(&input)), true),
    };

    #[cfg(not(miri))]
    {
        insta::assert_snapshot!(snapshot);
        let _ = failure;
    }

    #[cfg(miri)]
    {
        if failure {
            panic!("{snapshot}");
        } else {
            eprintln!("{snapshot}");
        }
    }
});
