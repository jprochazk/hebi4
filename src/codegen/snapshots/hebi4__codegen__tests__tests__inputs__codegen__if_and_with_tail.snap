---
source: src/codegen/tests.rs
expression: snapshot
---
SOURCE
let a = 0
let b = 1

if a < b and b >= a {
  return 0
} else {
  return 1
}


OK
.fn @main
  params 0
  stack 3
  .locals
    a = r1
    b = r2
  .code
   0  lsmi r1, 0
   1  lsmi r2, 1
   2  islt r1, r2
   3  jmp 7   ; to 10
   4  isge r2, r1
   5  jmp 5   ; to 10
   6  lsmi r0, 0
   7  ret
   8  lnil r0
   9  jmp 4   ; to 13
  10  lsmi r0, 1
  11  ret
  12  lnil r0
  13  ret
